book/src/linking_and_dividing.md

```diff
@@ -0,0 +1,337 @@
+# Linking and Dividing
+
+This section introduces everything thats needed to write JNI functions. Upon
+finishing this chapter, you'll have written a method that divides two integers
+using Rust.
+
+# Whats in NativeAPI?
+
+`NativeAPI` is a class that lives in `jnibookjava`, the Java side of the example
+code that accompanies this book. It uses `System.loadLibrary` to locate the
+`jnibookrs` lib, and specifies native methods using the `native` keyword. The
+native method called `verify_link` was used earlier to ensure that the project
+was setup correctly, with environment variables.
+
+For style reasons, we always add native methods to `NativeAPI`. It doesn't need
+to be that way, it's just a convention that we follow here.
+
+```java
+package com.github.jni_rs.jnibook;
+
+class NativeAPI {
+
+   private static final Throwable INIT_ERROR;
+
+   // The static block will be executed the first time the NativeAPI
+   // class is used.
+   static {
+       Throwable error = null;
+       try {
+           System.loadLibrary("jnibookrs");
+       } catch (Throwable t) {
+           error = t;
+       }
+       INIT_ERROR = error;
+   }
+
+   private NativeAPI() {
+       // Not instantiable
+   }
+
+   static native int verify_link();
+}
+```
+
+
+## Function Linking Rules
+
+JNI identifies the functions it should link to using the package path, class
+name, and method name by following [these
+rules](https://docs.oracle.com/en/java/javase/11/docs/specs/jni/design.html).
+Since `.` isn’t allowed in C function names, the Java function's package path's
+`.` is mapped to `_`. However, since `_` could appear in a function name, it
+needs a different representation. So `_` gets replaced by `_1` before replacing
+`.` with `_`. And finally, the function name has to be prefixed with `Java_`, to
+avoid conflicts with other languages that support similar naming schemes.
```

Comment:
Wouldn't it be better to suggest that users run `javac -h <path-to-headers>`
and copy the function declarations from the generated header file(s)?

E.g.,
```
javac -h . NativeAPI.java
cat NativeAPI.h

# Or with Gradle
./gradlew build
cat target/<our-path>/NativeAPI.h
```

It doesn't have to be configured separately from usual javac invocation,
see for example, https://github.com/exonum/exonum-java-binding/blob/master/exonum-java-binding/core/pom.xml#L184). It can be put into our Gradle template, so that in `target/native-headers`
there are up-to-date headers with correct function names.

javac reliably produces correct method names, that a user can copy and paste.
Even though the transformations below are described really well, and the examples
illustrate them well, if someone tries to follow them IRL, it may be tedious:
  - Copy/type the package name
  - Separately copy/type method name
  - Mangle as specified.

I think the rules distract from the more interesting parts; and may be harmful:
it is possible to mis-understand them, or make a typo or an error during a manual
conversion, or miss "Java_", and end up debugging "LinkageError: symbol not found".

Surely, it would be lovely to not have to do that at all: https://github.com/jni-rs/jni-rs/issues/81

Reply:
Decided to just use javac for now